### ISSUE TITLE
PriorityQueue is far from FIFO for same priority Pods

### ISSUE THREAD

**Issue by yqwang-ms** [kubernetes/kubernetes#83834]
**What happened**:
By default, user Pod does not have priority (i.e. 0), so usually, they expect **FIFO** scheduling in this case, even if for the PriorityQueue.
In this requirement, the **FI(First In)** should be determined by the [Pod CreationTimestamp instead of last queued time](https://github.com/kubernetes/kubernetes/issues/83826), i.e. the **user expect longest Pending Pod since creation should be scheduled first in most cases**.
However, we found that the PriorityQueue is far from FIFO, i.e. >40% Pods cut in line, see **How to reproduce it** for more details.

**What you expected to happen**:
The PriorityQueue should be **near FIFO** for same priority Pods, i.e. <5% Pods cut in line.

**How to reproduce it (as minimally and precisely as possible)**:
1. Cluster Setup: 7 Nodes each has 4 total free GPUs
2. Submit ReplicaSet RSX, which contains 28 Pods, and each Pod request 1 GPU and just sleep forever in its command.
3. Wait all Pods in RSX is running.
4. Submit Job JA, which contains 200 Pods to run to complete, and each Pod request 1 GPU and just sleep 20s in its command.
5. Wait all Pods in JA is queued in scheduler. (Because no free GPUs, so all Pods in JA are pending)
6. Submit Job JB, which has the same spec with JA except for its name JB is different with JA. (Here, we ensures JB is totally later than JA)
7. Delete ReplicaSet RSX, so JA and JB will compete to schedule.

We obeserves below result consistently:
| Timestamp (normalized) | JA completed Pods | JB completed Pods| Pods cut in line rate|
|:---- |:---- |:---- |:---- |
| 1 | 72 | 56 | 56/200 |
| 2 | 108 | 76 | 76/200 |
| 3 | 138 | 101 | 101/200 |
| 4 | 169 | 131 | 131/200 |
| 5 | 182 | 153 | 153/200 |
In avarage, >40% Pods cut in line.

**Anything else we need to know?**:
This issue can be fixed, i.e. cut in line rate decreased <5%, after we used the **Pod creation time instead of the last queued time to sort Pod for the same priority**, as suggested in https://github.com/kubernetes/kubernetes/issues/83826, and fixed by https://github.com/kubernetes/kubernetes/pull/83835.

After the fix https://github.com/kubernetes/kubernetes/pull/83835, we obeserves below result consistently (for the same repro steps above):
| Timestamp (normalized) | JA completed Pods | JB completed Pods| Pods cut in line rate|
|:---- |:---- |:---- |:---- |
| 1 | 160 | 1| 1/200 |
| 2 | 181 | 1| 1/200 |
| 3 | 191 | 3| 3/200 |
| 4 | 194 | 21| 21/200 |
In avarage, <5% Pods cut in line.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.4", GitCommit:"c27b913fddd1a6c480c229191a087698aa92f0b1", GitTreeState:"clean", BuildDate:"2019-02-28T13:37:52Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.2", GitCommit:"66049e3b21efe110454d67df4fa62b08ea79a19b", GitTreeState:"clean", BuildDate:"2019-05-16T16:14:56Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
```
- Scheduler Spec:
```
apiVersion: v1
kind: Pod
metadata:
  name: kube-scheduler
  namespace: kube-system
spec:
  hostNetwork: true
  containers:
  - image: gcr.io/google_containers/kube-scheduler:v1.14.2
    name: kube-scheduler
    command:
    - /usr/local/bin/kube-scheduler
    - --master
    - 10.151.41.16:8080
    - --leader-elect=false
    - --feature-gates=PodPriority=true
    - --v=4
    livenessProbe:
      httpGet:
        path: /healthz
        port: 10251
      initialDelaySeconds: 15
      timeoutSeconds: 1
    resources:
      requests:
        memory: "1Gi"
        cpu: "2000m"
```
- Cloud provider or hardware configuration:
`on-premise
`
- OS (e.g: `cat /etc/os-release`):
```
NAME="Ubuntu"
VERSION="16.04.6 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.6 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial

```
- Kernel (e.g. `uname -a`):
`Linux stcal-177 4.15.0-1060-azure #65-Ubuntu SMP Wed Sep 18 08:55:51 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
`
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by yqwang-ms** [comment]
/sig scheduling
@bsalamat and @denkensk PTAL

**Comment by hex108** [comment]
We came cross similar problem. 

> cut in line rate decreased <5%

How to guarantee it? Suppose that JA's pods are all in back off queue, and there are available GPU resource, then all JB's pods will be scheduled.

**Comment by yqwang-ms** [comment]
I cannot guarantee this, this is just a rough estimation for this specific reproduce steps.

> Suppose that JA's pods are all in back off queue, and there are available GPU resource, then all JB's pods will be scheduled.

Anyway, the probability for your case should be very small, because even if in current sort policy, JA and JB is **expected** to share the active queue evenly. And after #83835, JA is **expected** to contains more Pods in active queue than JB.

What user really care is the **expectation (average case)** of the cut in line rate instead of the **worst case** to guarantee, and #83835 's goal is to reduce the former.

BTW, it need some mathematical calculations to get the [expectation](https://en.wikipedia.org/wiki/Expected_value) of the cut in line rate, but we definitely know that it must be less than current sort policy.

**Comment by riking** [comment]
In your example, it seems like the cluster is getting less work done after your change?

| 4 | 169 | 131 | 131/200 |
versus
| 4 | 194 | 21| 21/200 |

300 jobs done > 215 jobs done, so your throughput is better before the change

Do jobs A and B have ordering requirements with each other?

**Comment by yqwang-ms** [comment]
@riking  Note the Timestamp in the table is normalized, which means one specific value only have relative meaning in its own table. From what I tested, the throughout looks the same before and after this change.
Job A and Job B doesn't depend on each other, this issue is just mentioned they are queued in an unfair way.

**Comment by fribourgpierre** [comment]
Hi,

I'm coming from #83835 about the same issue.

I am facing the same problem of old pods scheduled latter than new pods on a static cluster (see https://github.com/kubernetes/kubernetes/pull/83835#issuecomment-584576933)

@yqwang-ms suggest me to use #83835 PR but I may then face starvation issue mentioned in #86373. He also mention that I could leverage the queue sort plugin.

I didn't knew about any "sort plugin" before, nor how to use it but it seems to be the good approach, how can we use it?
Also, if anyone have other workaround, it would be very appreciated.

Thanks all.

**Comment by yqwang-ms** [comment]
right you need to write your own plugin by the link you provided

**Comment by alculquicondor** [comment]
Maybe this was asked before, but, if so, better resurface those conclusions in this issue.

Why should the scheduler be near FIFO? If a user comes in with a lot of jobs, they will almost completely starve other users that would come later. This is not like that in modern operating systems. Instead, I believe our current strategy tries to be more responsive to new jobs.

**Comment by fribourgpierre** [comment]
@alculquicondor it don't have to be like this, anyway we are searching a possibility to configure it this way for some use cases.
eg. a limited nb of available nodes and a large number of jobs coming in that needs to provide a answer as fast as possible regarding their arriving time.

**Comment by yqwang-ms** [comment]
Seems when no explicit defined policies to judge the priority of 2 tasks (such as priority field, user max tasks limit, etc), the FIFO rule can be the default policy which meets most human sense.

**Comment by alculquicondor** [comment]
Again, near FIFO is not how most operating systems work today. I don't think that would be the user expectation. I would like to know more arguments, given that you suggest that FIFO should be the default.

**Comment by yqwang-ms** [comment]
Pls check this https://github.com/kubernetes/kubernetes/pull/83835#discussion_r335299603

**Comment by alculquicondor** [comment]
That seems reasonable as an argument to have an alternate behavior, but not to make it the default.
Personally, I think FIFO is a greater source of starvation. And that is why operating systems schedulers moved away from it to be more responsive to all users. But if I understand correctly, the appeal is coming from the batch-type jobs. So maybe a better solution for it is co-scheduling, for which a KEP is in review: kubernetes/enhancements#1463

Possible courses of action:

- Work on a queue sort plugin on https://github.com/kubernetes-sigs/scheduler-plugins
- Write a KEP that would explain the motivation and use cases.

**Comment by yqwang-ms** [comment]
We have already solved the co-scheduling/gang-scheduling problem by scheduler extender.
However:
For job requires gang-scheduling, non-FIFO still impacts among these jobs (pod groups).
For job does not require gang-scheduling, non-FIFO not only impact among these jobs, but also impacts among their pods.

Besides, to be FIFO, does not have to sacrifice too much responsive or utilization, see my proposal in #86373.

**Comment by yqwang-ms** [comment]
BTW, if we do not respect FIFO, in theory, it can still cause starvation. Because the early pod may be scheduled later than later pod, so it may not be scheduled forever.
What I mean is that, it is not FIFO that cause starvation, it is current K8S **implementation** (priority queue + backoff queue), that to be FIFO caused starvation.

**Comment by alculquicondor** [comment]
> We have already solved the co-scheduling/gang-scheduling problem by scheduler extender.

Note that the KEP is also using the queuesort endpoint. Extenders can't do that.

> BTW, if we do not respect FIFO, in theory, it can still cause starvation. Because the early pod may be scheduled later than later pod, so it may not be scheduled forever.

An earlier pod will eventually have an earlier timestamp. Quoting Bobby again:

> The existing logic of sorting based on last schedule attempt timestamp is fair. If we sorted pods with the same priority by their creation time, an unschedulable pod created earlier than others would get more scheduling cycles than other pods with the same priority.

I'm trying to advocate for policy more than implementation. The implementation had issues with backoffs that we have fixed for 1.18, although there is still #87850 which is awaiting for experiments. I believe this is the last piece to solve HOL blocking.

> Because FIFO is equivalent to pods with different priorities, if latter doesn't have starvation anymore, former also doesn't.

This is true, but dangerous. We are basically saying that earlier pods have higher priority. Even if there is no HOL, responsiveness for new pods will be affected.

Other than that, I think your proposed algorithm in #86373 has one problem. As I understand, unschedulable pods are inserted back to the queue after each cycle. If there are too many unschedulable pods with higher priority or early creation time, the throughput of the scheduler is reduced. Because the scheduler has to retry those pods before moving to the newer pods that might fit.

But now we are talking about two problems in two different threads. I think we need a KEP or a doc with your proposal so that we can more easily comment on specific aspects.

**Comment by alculquicondor** [comment]
> I believe this is the last piece to solve HOL blocking.

Actually, it's not the last piece. We could try to be more diligent identifying the events that bring pods from the unschedulable queue back into the scheduling queue. But that is a bigger effort with too many corner cases, whereas #87850 is more low-hanging.

**Comment by yqwang-ms** [comment]
Extender can do gang scheduling, because in our setup, the pods in the same group shares the same priority (with job creation time encoded) with the whole job. So even without queue sort plugin, and with our gang reservation, all pods in the group will eventually (normally very fast in fact) fill to the reservation and get allocated.

BTW, I avoid to compare k8s with modern OS, because OS is so different than current k8s, for example, the preemption in OS, normally just suspend process, instead of killing it, while K8S preemption is killing. So OS can switch to different process without too much hurts, but k8s cannot. So k8s needs to take more care to select which pod to run now. If the later pod is selected, then the early pod may have to wait until it complete. If then another later pod2 is selected, early pod may also need to wait pod2. This is not intuitive for users.

**Comment by fribourgpierre** [comment]
Sorry to diverge from the root debate but does anyone of you can help implementing a workaround to create a FIFO QueueSort plugin on Scheduler Framework with out-of-tree feature (https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/20180409-scheduling-framework.md#custom-scheduler-plugins-out-of-tree) ?

As I am not familiar Scheduler Framework, it seems that I need to compile a full Scheduler, and that I have to write it in Golang only.
Documentation is very light on this topic, and I can't find any clear example on how to make & deploy such a plugin.
Does anyone have a link with a clear how-to about such task (out-of-tree technique) ?

Basically the less() function using K8S Go SDK looks easy to implement, it's more about deploying the plugin onto/with the Scheduler.

Thanks a lot for you help.

**Comment by alculquicondor** [comment]
The framework is quite new and we are only fully leveraging it in 1.18. In the future, we would like to have a sample in https://github.com/kubernetes-sigs/scheduler-plugins
That is indeed the doc to set it up https://kubernetes.io/docs/tasks/administer-cluster/configure-multiple-schedulers once you have the code. But better use the mailing list or slack channel if you have further questions.

**Comment by alculquicondor** [comment]
> I avoid to compare k8s with modern OS, because OS is so different than current k8s

The truth is that different users will always have a different view of their clusters. Following with the OS comparison, there are specialized schedulers for RT needs.

The current behavior follows a multi-user environment were we try to avoid a single deployment (as a concept, not necessarily a `apps/v1.Deployment`) from overtaking the cluster. And this is useful if several services are trying to scale up at roughly the same time. It seems that your view is more of a batch environment were a job should finish before another one is allowed to start. Unless there is enough space for both.

Changing a default behavior is not an easy task and it requires to follow release and deprecation policies. Users might be relying on it.

But, that doesn't prevent us to add new optional behaviors. And this is simpler with the framework. We just need the KEP :)

**Comment by yqwang-ms** [comment]
Agree! It is hard or impossible to make single scheduler sutiable for all scenarios.
BTW, for the first step, we can clarify current scheduler behavior, such as not FIFO, in our wiki:
https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#effect-of-pod-priority-on-scheduling-order

**Comment by alculquicondor** [comment]
Feel free to send a PR to kubernetes/website. And perhaps introduce a plugin in kubernetes-sigs/scheduler-plugins

**Comment by alculquicondor** [comment]
/remove-lifecycle stale

cc @adtac for future reference

**Comment by alejandrox1** [comment]
/remove-lifecycle stale
/cc @kubernetes/sig-scheduling-bugs @kubernetes/sig-scheduling-feature-requests 
ptal

**Comment by Huang-Wei** [comment]
FYI: there is a [KEP](https://github.com/kubernetes/enhancements/pull/2781) proposing surfacing the whole SchedulerQueue interface to be implementable:

**Comment by jasonliu747** [comment]
@yqwang-ms Is there any progress fixing this issue using Scheduling Framework?
